### PR TITLE
Fixed #27113 -- Added a test for setting HttpRequest.encoding.

### PR DESCRIPTION
**Pull request by PREM1980**
https://code.djangoproject.com/ticket/27113

### CODE REVIEW

**Comment by PREM1980** [comment]
Fixed 27113.

File: tests/requests/tests.py [hunk 1]
```diff
@@ -872,3 +872,13 @@ def test_build_absolute_uri_relative_location(self):
             request.build_absolute_uri(location='/foo/bar/'),
             'http://testserver/foo/bar/'
         )
+
+    def test_set_encoding(self):
```
Comment (timgraham) [inline tests/requests/tests.py]: This doesn't look appropriate for `BuildAbsoluteURITestCase`.

File: tests/requests/tests.py [hunk 2]
```diff
@@ -872,3 +872,13 @@ def test_build_absolute_uri_relative_location(self):
...
+        payload = FakePayload('name=Hello Günter')
+        request = WSGIRequest({'REQUEST_METHOD': 'POST',
```
Comment (timgraham) [inline tests/requests/tests.py]: Use hanging indent:
```
request = WSGIRequest({
    'REQUEST_METHOD': 'POST',
    ...
})

**Comment by timgraham** [comment]
How about these lines: https://github.com/django/django/blob/bc1e2d8e8edde6cc7d2657c68242a13ee65a15b8/django/http/request.py#L228-L229

I think they might be unused.

**Comment by PREM1980** [comment]
Fixed everything.
Here is the new code.
The POST works fine but the GET does not encode the string. The tests fails.

Not sure if I need to open a new ticket to check if the encoding works or not for GET.


def test_set_encoding(self):
        payload = FakePayload('name=Hello Günter')
        request = WSGIRequest({'REQUEST_METHOD': 'POST',
                               'CONTENT_TYPE': 'application/x-www-form-urlencoded',
                               'CONTENT_LENGTH': len(payload),
                               'wsgi.input': payload})
        self.assertEqual(request.POST, {'name': ['Hello Günter']})
        request.encoding = 'iso-8859-16'
        self.assertEqual(request.POST, {'name': ['Hello GĂŒnter']})
        # Test for GET        
        request = WSGIRequest({
            'REQUEST_METHOD': 'GET',
            'wsgi.input': BytesIO(b''),
            'QUERY_STRING': 'name=Hello Günter'
        })
        request.encoding = 'iso-8859-16'        
        self.assertEqual(request.GET, {'name': ['Hello GĂŒnter']})




Here is the failure 👎 
FAIL: test_set_encoding (requests.tests.RequestsTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/unittest/case.py", line 58, in testPartExecutor
    yield
  File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/unittest/case.py", line 600, in run
    testMethod()
  File "/Users/premanandlakshmanan/Documents/django-contribute/django/tests/requests/tests.py", line 593, in test_set_encoding
    self.assertEqual(request.GET, {'name': ['Hello GĂŒnter']})
  File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/unittest/case.py", line 820, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/unittest/case.py", line 813, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: <QueryDict: {'name': ['Hello Günter']}> != {'name': ['Hello GĂŒnter']}

**Comment by timgraham** [comment]
Looks like the commit that broke `GET` is d7a4b156d935ffcb8cdd9290688457453fc06b24. I think the line in the `encoding.settr` needs to use `del self.GET` instead ([docs about clearing cached_property](https://docs.djangoproject.com/en/1.9/ref/utils/#django.utils.functional.cached_property)).

**Comment by timgraham** [comment]
By the way, you can make the changes for GET in a separate commit since it's actually fixing a bug. In fact, it would be appropriate to open a separate ticket.

**Comment by PREM1980** [comment]
Ok, great, should I submit my patch only for testing POST request.

**Comment by timgraham** [comment]
Sure

**Pull request closed without merging**